src/backend/auxil/range_alloc.rs

```diff
@@ -0,0 +1,169 @@
+    /// No two ranges in this vec may overlap.
+    empty_ranges: Vec<Range<T>>,
+}
+
+impl<T> RangeAllocator<T>
+    where T: Clone + Copy + Add<Output=T> + AddAssign + Sub<Output=T> + Eq + PartialOrd + Debug
+{
+    pub fn new(range: Range<T>) -> Self {
+        RangeAllocator {
+            range: range.clone(),
+            empty_ranges: vec![range.clone()],
+        }
+    }
+
+    pub fn range(&self) -> Range<T> {
+        self.range.clone()
+    }
+
+    pub fn allocate_range(&mut self, length: T) -> Option<Range<T>> {
+        let mut best_range_tuple: Option<(usize, Range<T>)> = None;
+        for range in self.empty_ranges.iter().cloned().enumerate() {
+            if range.1.end - range.1.start < length {
+                continue;
+            }
+            match best_range_tuple {
+                Some((ref mut index, ref mut best_range)) => {
+                    // We want to find the best fit for this allocation in order to reduce 
+                    // memory fragmentation, so we'll look for it here.
+                    if range.1.end - range.1.start < best_range.end - best_range.start {
+                        *index = range.0;
+                        *best_range = range.1.clone();
+                        // Found a perfect fit, so stop looking.
+                        if range.1.end - range.1.start == length {
+                            break;
+                        }
+                    }
+                },
+                None => { 
+                    best_range_tuple = Some(range.clone());
+                    // Found a perfect fit, so stop looking.
+                    if range.1.end - range.1.start == length {
+                        break;
+                    }
+                }
+            }
+        }
+        best_range_tuple.map(|range| {
+            if range.1.end - range.1.start == length {
+                self.empty_ranges.remove(range.0);
+            } else {
+                self.empty_ranges[range.0].start += length;
+            }
+            range.1.start..(range.1.start + length)
+        })
+    }
+
+    pub fn free_range(&mut self, range: Range<T>) -> bool {
+        assert!(range.start >= self.range.start && range.end <= self.range.end);
+        assert!(range.start <= range.end);
+        if self.empty_ranges.len() == 0 {
+            self.empty_ranges.push(range);
+            return true;
+        }
+        // Input is within range, but before any empty ranges
+        if self.empty_ranges.len() > 0 {
+            if self.empty_ranges[0].start == range.end {
+                self.empty_ranges[0].start = range.start;
+                return true;
+            } else if self.empty_ranges[0].start > range.end {
+                self.empty_ranges.insert(0, range);
+                return true;
+            }
+        }
+        // Input is within range, but after all empty ranges
+        if let Some(last) = self.empty_ranges.last().cloned() {
```

Comment:
Actually I don't think all these early out special cases are not really needed. Basically hoping the users follow some sort of pattern when freeing but I don't believe that's the case.

Reply:
This comment is confusing because of a few double negatives.  Can you reword it?

Reply:
sry I meant 'early out special cases are not really needed'.
Everything is handled in the iteration case as far as I can see.

Reply:
I thought so too until I added some unit tests.  I added these because the iteration wasn't accounting for these cases.